LORE-96 Create initial version of the service registry

mdang8 opened the pull request:
The initial version of the service registry listens to the transform request queue and moves the requests to the service queue for further processing. The next version will have the database lookup functionalities using Spring Data JPA. The Docker image is named `transformation-service-registry` and can be created by running the `build.sh` script. The Docker stack currently has the service registry and a RabbitMQ instance as services on the same network. The stack can be deployed by running `docker stack deploy -c docker-compose.yml trs` with `trs` being the name of the Docker stack.

## Reviewers

@josephthweatt @cjlange @figliold @mcalcote @brendan-hofmann

mdang8 committed:
Add AMQP configs and queue listener/producer implementations

josephthweatt commented on build.sh:
```diff
@@ -0,0 +1,6 @@
+#!/bin/bash
```
Maybe change to build-image, or something to specify that this is for creating an image

mdang8 replied:
Sure, I can do that.

mdang8 committed:
Rename build script

josephthweatt commented on src/main/java/com/connexta/transformation/api/ServiceRegistryQueueConstants.java:
```diff
@@ -0,0 +1,24 @@
+  public static final String REQUEST_ROUTING_KEY = "request";
+  public static final String SERVICE_ROUTING_KEY = "sodars";
+
+  private ServiceRegistryQueueConstants() {
```
Do we need this method for anything?

mdang8 replied:
I just put that private constructor so Sonar Lint would stop complaining.

josephthweatt commented on src/main/java/com/connexta/transformation/api/ServiceRegistryQueueConstants.java:
```diff
@@ -0,0 +1,24 @@
+ */
+package com.connexta.transformation.api;
+
+public abstract class ServiceRegistryQueueConstants {
```
This class is never extended, should consider whether to keep it an abstract class

mdang8 replied:
Actually, I don't remember why I made that abstract... In any case, I don't really like having the class for constants. I'm going to look into just being able to throw those values in the application properties file.

mdang8 committed:
LORE-96 Remove queue constants class

figliold commented on Dockerfile:
```diff
@@ -0,0 +1,21 @@
+
+FROM openjdk:8-jre-alpine
+VOLUME /tmp
+WORKDIR /transformation-app
```
We probably shouldn't hardcode the version in the Dockerfile.

brendan-hofmann replied:
Should or should not? I would think the dockerfile should be a filtered resource so it can use the version from the pom.

mdang8 replied:
I removed the hardcoded version and instead pass it in as an argument from the fabric8 Maven plugin.

figliold commented on pom.xml:
```diff
@@ -0,0 +1,101 @@
+  </properties>
+
+  <build>
+    <plugins>
```
We should add the spotless plugin for code formatting.

brendan-hofmann replied:
We also need OWASP and SonarQube.

figliold commented on transformation-service-registry.iml:
```diff
@@ -0,0 +1,82 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
.iml files should be in the .gitignore file.

mdang8 replied:
Fixed.

figliold commented on pom.xml:
```diff
@@ -0,0 +1,101 @@
+  <dependencies>
+    <dependency>
+      <groupId>com.rabbitmq</groupId>
+      <artifactId>amqp-client</artifactId>
```
Should we move all of the dependency versions into maven properties?

brendan-hofmann replied:
And/or dependency-management.

mdang8 replied:
I put the dependency versions in a `<properties>` tag.

figliold commented on pom.xml:
```diff
@@ -0,0 +1,101 @@
+  </parent>
+
+  <dependencies>
+    <dependency>
```
Do you need this dependency or will the already spring-rabbit and spring-amqp work (both of those are already included as dependencies).

mdang8 replied:
No those aren't necessary. I removed them.

figliold commented on pom.xml:
```diff
@@ -0,0 +1,101 @@
+  <groupId>transformation</groupId>
+  <artifactId>ion-transformation-service-registry</artifactId>
+  <version>0.0.1-SNAPSHOT</version>
+
```
I don't think we want this as the parent.

figliold commented on pom.xml:
```diff
@@ -0,0 +1,101 @@
+      <groupId>com.jayway.jsonpath</groupId>
+      <artifactId>json-path</artifactId>
+      <scope>test</scope>
+    </dependency>
```
Is this dependency needed?

mdang8 replied:
No, I removed them.

figliold commented on src/main/java/com/connexta/transformation/AmqpConfiguration.java:
```diff
@@ -0,0 +1,130 @@
+
+  @Value("${spring.rabbitmq.request-queue-name")
+  private String requestQueueName;
+
```
For now this is okay, but the service queue will eventually be looked up and will depend on the product type.

figliold commented on src/main/java/com/connexta/transformation/Application.java:
```diff
@@ -0,0 +1,32 @@
+
+  @Override
+  public void run(String... args) throws Exception {
+
```
You should remove the System.out.println() and the sleep.

brendan-hofmann replied:
Should we be using something like SLF4J for logging?

mdang8 replied:
I have SLF4J as a dependency and removed the sleep function.

figliold commented on src/main/java/com/connexta/transformation/Application.java:
```diff
@@ -0,0 +1,32 @@
+import org.springframework.boot.SpringApplication;
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+
+@SpringBootApplication
```
You shouldn't need to implement CommandLineRunner. This should work:
public static void main(String[] args) {
    SpringApplication.run(Application.class, args);
  }

mdang8 replied:
Fixed.

figliold commented on .gitignore:
```diff
@@ -0,0 +1,6 @@
+.DS_Store
```
You may want to add a few more things to this: See https://github.com/connexta/discovery-api/blob/master/.gitignore

figliold commented on src/main/java/com/connexta/transformation/AmqpConfiguration.java:
```diff
@@ -0,0 +1,130 @@
+
+  @Bean
+  public MessageListenerAdapter listenerAdapter() {
+    ServiceRegistryConsumer consumer = new ServiceRegistryConsumerImpl();
```
Make consumeFromQueue a constant.

mdang8 replied:
Done.

figliold commented on src/main/java/com/connexta/transformation/api/ServiceRegistryConsumer.java:
```diff
@@ -0,0 +1,18 @@
+ */
+package com.connexta.transformation.api;
+
+public interface ServiceRegistryConsumer {
```
Does this only throw an InterruptedException because you have a temporary sleep it the impl?  If so, we should probably remove it from the interface.

mdang8 replied:
Yeah that was the only reason. Removed it.

figliold commented on src/main/java/com/connexta/transformation/impl/ServiceRegistryConsumerImpl.java:
```diff
@@ -0,0 +1,43 @@
+  }
+
+  @Override
+  public void consumeFromQueue(byte[] transformRequest) throws InterruptedException {
```
I wonder if you can use the Jackson2JsonMessageConverter (https://docs.spring.io/spring-amqp/api/org/springframework/amqp/support/converter/Jackson2JsonMessageConverter.html) to do all of the json serialization and deserialization?

mdang8 replied:
Yeah it looks like I can. Replaced it with that.

figliold commented on src/main/java/com/connexta/transformation/impl/ServiceRegistryProducerImpl.java:
```diff
@@ -0,0 +1,38 @@
+@Component
+public class ServiceRegistryProducerImpl implements ServiceRegistryProducer {
+  private final ApplicationContext context = new AnnotationConfigApplicationContext(
+      AmqpConfiguration.class);
```
You should be able to inject the AmqpTemplate instead of doing this lookup from the context.

mdang8 replied:
Fixed.

figliold commented on src/main/java/com/connexta/transformation/impl/ServiceRegistryConsumerImpl.java:
```diff
@@ -0,0 +1,43 @@
+
+    // @TODO - do work (query database)
+    Thread.sleep(5000);  // @TODO - remove this
+
```
Do we really need to create a new producer for each publish to the queue?

mdang8 replied:
No, I don't. I added an initialization of a single publisher for each instance of the consumer.

cjlange commented on src/main/java/com/connexta/transformation/AmqpConfiguration.java:
```diff
@@ -0,0 +1,130 @@
+import com.connexta.transformation.api.ServiceRegistryConsumer;
+import com.connexta.transformation.impl.ServiceRegistryConsumerImpl;
+
+@Configuration
```
@mdang8 - ✏️ Add a *JavaDoc* here to clarify what this Config is used for.

mdang8 replied:
Done.

cjlange commented on src/main/java/com/connexta/transformation/api/ServiceRegistryConsumer.java:
```diff
@@ -0,0 +1,18 @@
+ */
+package com.connexta.transformation.api;
+
+public interface ServiceRegistryConsumer {
```
@mdang8 - ✏️ Could you add a *JavaDoc*?

mdang8 replied:
Done.

cjlange commented on src/main/java/com/connexta/transformation/api/ServiceRegistryProducer.java:
```diff
@@ -0,0 +1,5 @@
+package com.connexta.transformation.api;
+
+public interface ServiceRegistryProducer {
```
@mdang8 - ✏️ Could you add a *JavaDoc*?

mdang8 replied:
Done.

cjlange commented on src/main/java/com/connexta/transformation/impl/ServiceRegistryConsumerImpl.java:
```diff
@@ -0,0 +1,43 @@
+import com.connexta.transformation.api.ServiceRegistryProducer;
+
+@Component
+public class ServiceRegistryConsumerImpl implements ServiceRegistryConsumer {
```
@mdang8 - ✏️ Could you add a *JavaDoc*?

mdang8 replied:
Done.

cjlange commented on src/main/java/com/connexta/transformation/impl/ServiceRegistryConsumerImpl.java:
```diff
@@ -0,0 +1,43 @@
+public class ServiceRegistryConsumerImpl implements ServiceRegistryConsumer {
+  private final Logger LOGGER = LoggerFactory.getLogger(ServiceRegistryConsumerImpl.class);
+
+  public ServiceRegistryConsumerImpl() {
```
@mdang8 - ✏️ Do we need this constructor for something in the future?

mdang8 replied:
I changed it to inject the AMQP template in the constructor.

cjlange commented on src/main/java/com/connexta/transformation/impl/ServiceRegistryConsumerImpl.java:
```diff
@@ -0,0 +1,43 @@
+    LOGGER.info(String.format("Querying for mime type '%s'", mimeType));
+
+    // @TODO - do work (query database)
+    Thread.sleep(5000);  // @TODO - remove this
```
@mdang8 - ✏️ Is there a ticket you could reference for this *TODO* work?

mdang8 replied:
There isn't one that I can currently use.

cjlange commented on src/main/java/com/connexta/transformation/impl/ServiceRegistryProducerImpl.java:
```diff
@@ -0,0 +1,38 @@
+      AmqpConfiguration.class);
+  private final AmqpTemplate amqpTemplate = context.getBean(RabbitTemplate.class);
+
+  public ServiceRegistryProducerImpl() {
```
@mdang8 - ✏️ Do we need this constructor for something in the future?

mdang8 replied:
I changed it to inject the AMQP template in the constructor.

cjlange commented on src/main/java/com/connexta/transformation/impl/ServiceRegistryProducerImpl.java:
```diff
@@ -0,0 +1,38 @@
+
+  @Override
+  public void publishToQueue(byte[] transformRequest) {
+    this.amqpTemplate.convertAndSend(transformRequest);
```
@mdang8 - `this` is not required here.

mdang8 replied:
It's not required, but it's just a personal preference of mine to make it easier to distinguish. I don't know if we have any standards for it.

mdang8 replied:
I decided to change it and removed the usages of the `this` keyword to follow the code style of the rest of the team.

cjlange reviewed (changes requested):
@mdang8 - A few minor comments.

josephthweatt commented:
Also might be good to add a PR template like we have in DDF: https://github.com/codice/ddf/blob/master/.github/PULL_REQUEST_TEMPLATE

brendan-hofmann commented on build-docker.sh:
```diff
@@ -0,0 +1,6 @@
+#!/bin/bash
```
❗️ We should be using a docker plugin to build the image as part of the maven build so CI can build it. Consider the fabric8 docker plugin: https://dmp.fabric8.io

mdang8 replied:
Added it to the POM and removed the build script.

mdang8 committed:
LORE-96 Move application properties to environment variables

mdang8 committed:
LORE-96 Address some PR reviews

mdang8 committed:
LORE-96 Fix application properties constants

mdang8 committed:
LORE-96 Add Javadocs and remove use of config context declaration

mdang8 committed:
LORE-96 Fix POM file dependencies for PR reviews

mdang8 committed:
LORE-96 Replace build script with fabric8 Docker plugin

mdang8 committed:
LORE-96 Rename JAR file used in Docker container

mdang8 committed:
LORE-96 Ignore .iml files

mdang8 committed:
LORE-96 Remove hardcoded version from Dockerfile

Pass in argument from fabric8 Maven plugin

mdang8 committed:
LORE-96 Address final PR reviews

Add more Javadocs, remove usage of `this` keyword, and use Jackson library for message conversion

brendan-hofmann commented:
Should we add a `.gitattributes` to enforce line endings?

brendan-hofmann commented on pom.xml:
```diff
@@ -0,0 +1,281 @@
+        <artifactId>dependency-check-maven</artifactId>
+        <version>${dependency-check-maven-plugin.version}</version>
+        <configuration>
+          <!-- @TODO Decide how to handle dependency check failures -->
```
Should we have some tracking, such as a ticket, for this TODO?
